MINOR: Code Cleanup

vahidhashemian opened the pull request:
Clean up includes:

- Switching try-catch-finally blocks to try-with-resources when possible
- Removing some seemingly unnecessary `@SuppressWarnings` annotations
- Resolving some Java warnings
- Closing unclosed Closable objects
- Removing unused code

vahidhashemian committed:
MINOR: Code Cleanup

Clean up includes
- Removing some unused `@SuppressWarnings` annotations
- Resolved some Java warnings
- Closing unclosed Closable objects
- Removing unused code

baluchicken commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -85,16 +85,17 @@ public void testConstructorFailureCloseResource() {
         final int oldInitCount = MockMetricsReporter.INIT_COUNT.get();
         final int oldCloseCount = MockMetricsReporter.CLOSE_COUNT.get();
+        KafkaProducer<byte[], byte[]> producer = null;
         try {
```
Why shouldn't we use the try with resources java 7 feature here?

vahidhashemian replied:
Thanks. That's a great suggestion. I'll update the patch to use it.

vahidhashemian committed:
Switched try-catch-finally blocks to try-with-resources when possible

vahidhashemian commented:
cc @ijuma for feedback. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5360/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5345/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:
```diff
@@ -77,6 +76,7 @@ public void testPartitioner() throws Exception {
         assertEquals("Partition should be correct", 1, metadata.get().partition());
         producer.clear();
         assertEquals("Clear should erase our history", 0, producer.history().size());
+        producer.close();
```
It's odd that we close this producer, but not the one assigned to the field with the same name. Do we care about closing `MockProducer`? Did you use a tool to find cases where `Closeable`s are not being closed?

vahidhashemian replied:
Thanks for catching this. I think I missed some `close()` calls on the member variable. I basically looked for usages of `Closable` objects (did not use a tool, so I may have missed some). For this particular one, `MockProducer` implements the `Producer` (which is `Closable`). And it looks like I need to close both the local and the member variables. I'll add that in the next commit.

I think I'll add the `close()` for member variable in a junit `@After` method so the producer is also closed when we are expecting exceptions. That means in some methods that `close()` is called earlier for the test we need to define additional local variables.

I hope that makes sense. Please advise if you can think of a better approach. Thanks.

mjsax replied:
Why don't we use try-with-resource pattern?

vahidhashemian replied:
For this PR I tried to use try-with-resource pattern instead of an existing try-catch block, when possible. I'm not sure about the best-practice here, but my feeling is the code is more readable this way, plus I'm not sure if try-with-resource adds much value here. Please advise if you disagree.

mjsax replied:
Don't think readability would suffer. For `MockProducer` is might also not be bad atm if `close()` is not called in case the test fails with an exception and thus `close()` is never called. It just feel "right" to make sure to call `close()` for all classes that are `Closable` -- even in case of failure. This might also be more "future prove" in case `close()` implementation changes at some point. But I also don't insist on the change.

vahidhashemian replied:
Sure, that sounds good. I think this pattern could be applied in other places (files) too, but I'll skip them for now. I'll submit a new commit. Please let me know if you still see room for improvement. Thanks.

ijuma commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java:
```diff
@@ -72,8 +72,10 @@ private void load() {
         try {
             ObjectInputStream is = new ObjectInputStream(new FileInputStream(file));
```
Why not try with resources?

vahidhashemian replied:
Yes, I had missed this one. Thanks.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java:
```diff
@@ -372,8 +372,10 @@ public void testConsume(final long prodTimeMs) throws Throwable {
         final List<PartitionInfo> partitionInfos = consumer.partitionsFor(testConfig.topic);
-        if (partitionInfos.size() < 1)
+        if (partitionInfos.size() < 1) {
+            consumer.close();
```
Maybe easier to use try with resources?

vahidhashemian replied:
This one seems a bit non-trivial to use try with resources. I'll give it a shot :)

ijuma reviewed (commented):
Thanks for cleaning up the code. Left some feedback. We'd also want someone more familiar with Streams to take a look at the Streams part. cc @enothereska @guozhangwang

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -223,7 +223,7 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
     }
 
-    @SuppressWarnings({"unchecked", "deprecation"})
+    @SuppressWarnings("unchecked")
```
Hmm.. I'm wondering why it was added at the first place :P

vahidhashemian replied:
It was added a long time ago by [this PR](https://github.com/apache/kafka/commit/241c3ebb2803f1e09306fb06f20a66e7a60ca3c8), due to deprecated producer configs, such as `ProducerConfig.BLOCK_ON_BUFFER_FULL_CONFIG`, that no longer exist.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -53,7 +53,6 @@
  * @see KGroupedStream
  * @see KStreamBuilder#stream(String...)
  */
-@SuppressWarnings("unused")
```
This is used as some of the apis are never used within the library itself, curious to learn what's the motivation of removing it.

vahidhashemian replied:
My IDE identified this line as unnecessary. I guess it doesn't check whether the APIs are used or not. I can put it back if you think it should be there.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -58,7 +58,6 @@
     private final Map<TopicPartition, Long> checkpointedOffsets;
     private final Map<String, StateRestoreCallback> restoreCallbacks; // used for standby tasks, keyed by state topic name
     private final Map<String, String> storeToChangelogTopic;
-    private final boolean eosEnabled;
```
cc @mjsax

mjsax replied:
Variable is unused so that's ok. Seems to be slipped to remove it after some code refactoring (cannot recall anymore why I introduced it originally).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:
```diff
@@ -128,6 +128,7 @@ public static AssignmentInfo decode(ByteBuffer data) {
             if (version < 0 || version > CURRENT_VERSION) {
                 TaskAssignmentException ex = new TaskAssignmentException("Unknown assignment data version: " + version);
                 log.error(ex.getMessage(), ex);
+                in.close();
```
Should we also close the input stream at line 157 before returning as well?

vahidhashemian replied:
Yes, actually I'll try to use try with resources here in the next update. Thanks for catching it.

guozhangwang reviewed (commented):
lgtm. also cc'ing @mjsax for another look.

vahidhashemian committed:
Addressed feedback received, and did some additional refactoring

vahidhashemian reviewed (commented):
@ijuma @guozhangwang Thanks for reviewing the PR. I tried to address your concerns in the new commit. Please also see my responses inline, and advise if you still see issues. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5405/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5420/
Test PASSed (JDK 7 and Scala 2.11).

vahidhashemian commented:
@ijuma @guozhangwang @mjsax Thanks for your feedback on this PR. Do you see anything that still needs to be addressed? Thanks.

mjsax commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -369,7 +368,7 @@ public void testHeaders() throws Exception {
 
         String topic = "topic";
-        Collection<Node> nodes = Collections.singletonList(new Node(0, "host1", 1000));
+        Collections.singletonList(new Node(0, "host1", 1000));
```
It seem this can be removed completely? Compare line 375 (I guess, `nodes` was use there at some point but not anymore)

vahidhashemian replied:
That's correct. I'll remove the line in the next update. Thanks.

vahidhashemian reviewed (commented):
@mjsax Thanks for taking another look. Please see my responses inline. I'll send an update shortly to also resolve the conflict.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5674/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5660/
Test PASSed (JDK 8 and Scala 2.12).

vahidhashemian committed:
Added more try-with-resources blocks

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5689/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5675/
Test PASSed (JDK 8 and Scala 2.12).

vahidhashemian commented:
Any other feedback on this PR?
cc @hachikuji

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:
```diff
@@ -51,9 +52,12 @@
 
+    @After
+    public void close() {
+        producer.close();
```
Since we seem to be replacing many of the global references below, do we still need this? Perhaps we should just replace them all? Or perhaps we should use the "registration" approach that we use in some of the scala tests in order to make cleanup consistent between all cases?

vahidhashemian replied:
I agree there is some inconsistency as a result of introducing `try-with-resources` blocks. Even though quite a few global references are replaced, there are still a lot of test methods that use global references. I could introduce `try-with-resources` to all methods to make it consistent and get rid of the `@After` method above. By "registration" did you mean something like producers and consumers setup and teardown in [`IntegrationTestHarness.scala`](https://github.com/apache/kafka/blob/0bede30ada6d5719950ad35a69c586f0ba9c4f7e/core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala)?

hachikuji replied:
Yes, that's what I was thinking. Then we can remove the global references and use the same pattern consistently.

vahidhashemian replied:
I submitted an update, in which I removed the global variable and used `try-with-resources` blocks when it was possible. The tests now seem to have a consistent pattern.  I hope I did not misunderstand your comment.

hachikuji commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java:
```diff
@@ -69,19 +69,19 @@ public synchronized void stop() {
             Object obj = is.readObject();
-            if (!(obj instanceof HashMap))
+            if (!(obj instanceof HashMap)) {
+                is.close();
```
Why is this needed? Doesn't the `try` ensure that the input stream will be closed?

vahidhashemian replied:
Yes, you are right, as `close()` is called even when throwing the exception. Thanks for catching it.

hachikuji commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java:
```diff
@@ -74,6 +79,7 @@ public void castWholeRecordKeyWithSchema() {
 
         assertEquals(Schema.Type.INT8, transformed.keySchema().type());
         assertEquals((byte) 42, transformed.key());
+        xform.close();
```
Any reason we don't use the "try with resources" in these other cases as well? Maybe since the constructor has no arguments, we can just use a global instance and close in an `@After` method? Same for the other transforms below.

vahidhashemian replied:
Sure, I'll do that in the next patch, but I think in some of them we run into the same issue as the `MockProducerTest` issue you identified above. I hope I did not misunderstand your point.

hachikuji reviewed (commented):
Thanks for the patch. Left a few minor comments.

vahidhashemian reviewed (commented):
@hachikuji Thanks for your feedback. I'll try to submit an update to address two of your comments, while wait for your clarification on the third one.

vahidhashemian committed:
Addressed feedback by @hachikuji

- Removed a redundant `.close()` call.
- Made `xform` a global variable and replaced instances of `xform.close()` by and `@After` method

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6036/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6051/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java:
```diff
@@ -34,42 +35,54 @@
     public void testConfigEmpty() {
-        final Cast<SourceRecord> xform = new Cast.Key<>();
-        xform.configure(Collections.singletonMap(Cast.SPEC_CONFIG, ""));
+        try (final Cast<SourceRecord> xform = new Cast.Key<>()) {
```
We could instead use `xformKey`? Same for others below.

vahidhashemian committed:
Did additional refactoring

- Removed the global producer variable from `MockProducerTest`
- Reused the global `xform...` variable in `transforms` tests when possible

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6087/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6072/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:
```diff
@@ -47,22 +47,21 @@
-        assertEquals("We should have the record in our history", singletonList(record1), producer.history());
-        producer.clear();
-        assertEquals("Clear should erase our history", 0, producer.history().size());
+        try (MockProducer<byte[], byte[]> producer = new MockProducer<>(true, new MockSerializer(), new MockSerializer())) {
```
Have to say it's a bit annoying to have all these diffs just to close the mock producer (which does not have any resources to close). I wonder if we can find a lighter pattern. What I was trying to suggest before was to continue to use the shared `producer` with an `@After`. We can use a factory method to create it in each method. For example:
```java
@After
public void cleanup() {
  if (this.producer != null) this.producer.close()
}

private void buildMockProducer(boolean autoComplete) {
  this.producer = new MockProducer<>(autocomplete, new MockSerializer(), new MockSerializer()))
}
```
We can add overloads to `buildMockProducer` as needed. What do you think?

vahidhashemian replied:
Yup, this works too as another way of bringing consistency into these test. Sorry for misunderstanding your suggestion earlier. I'll submit an update shortly.

vahidhashemian committed:
Used a factory method to create the `MockProducer` object in `MockProducerTest`

vahidhashemian commented on clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:
```diff
@@ -70,185 +81,170 @@ public void testPartitioner() throws Exception {
-            producer.clear();
-            assertEquals("Clear should erase our history", 0, producer.history().size());
-        }
+        MockProducer<String, String> producer = new MockProducer<>(cluster, true, new DefaultPartitioner(), new StringSerializer(), new StringSerializer());
```
@hachikuji The only anomaly is here, where the producer type is different. Not sure if there is a cleaner way to handle it.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6136/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6152/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6137/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6153/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji reviewed (approved):
LGTM
